spirv_glsl.cpp

```diff
@@ -3583,12 +3583,22 @@ void CompilerGLSL::emit_extension_workarounds()
 		}
 		else if (shader_subgroup_supporter.is_feature_requested(Supp::SubgroupMemBarrier))
 		{
-			statement("#ifndef GL_KHR_shader_subgroup_basic");
-			statement("void subgroupMemoryBarrier() { groupMemoryBarrier(); }");
-			statement("void subgroupMemoryBarrierBuffer() { memoryBarrierBuffer(); }");
-			statement("void subgroupMemoryBarrierShared() { memoryBarrierShared(); }");
-			statement("void subgroupMemoryBarrierImage() { memoryBarrierImage(); }");
-			statement("#endif");
+			if (model == spv::ExecutionModelTessellationEvaluation || model == spv::ExecutionModelGLCompute)
+			{
+				statement("#ifndef GL_KHR_shader_subgroup_basic");
+				statement("void subgroupMemoryBarrier() { memoryBarrier(); }");
+				statement("void subgroupMemoryBarrierBuffer() { memoryBarrierBuffer(); }");
+				statement("void subgroupMemoryBarrierImage() { memoryBarrierImage(); }");
+				statement("#endif");
+			}
+			else
+			{
+				statement("#error Subgroup memory barriers can be emulated only in tesselation control and compute "
+				          "execution models.");
```

Comment:
this is not what I meant.

Reply:
I meant

### For Compute or Evaluation
```cpp
			statement("void subgroupMemoryBarrier() { groupMemoryBarrier(); }");
			statement("void subgroupMemoryBarrierBuffer() { groupMemoryBarrier(); }");
			statement("void subgroupMemoryBarrierShared() { groupMemoryBarrier(); }");
			statement("void subgroupMemoryBarrierImage() { groupMemoryBarrier(); }");
```
### For Everything Else
```cpp
			statement("void subgroupMemoryBarrier() { memoryBarrier(); }");
			statement("void subgroupMemoryBarrierBuffer() { memoryBarrierBuffer(); }");
			statement("void subgroupMemoryBarrierImage() { memoryBarrierImage(); }");
```

Reply:
blergh I made a mistake, its the Control shader not the Evaluation shader that's like compute

Reply:
and forgot that TCS can only use `barrier` but not shared mem and doesnt have explicit workgroups